File: md_prototype.cpp
#include <iostream>
#include <iomanip>
//#include <yaml-cpp/yaml.h>
#include "md_prototype.hpp"
#include "yaml/Yaml.hpp"

CoordReceiverAlgorithm::CoordReceiverAlgorithm(MPI_Comm comm, int rank, int partner_rank)
    : comm_(comm), rank_(rank), partner_rank_(partner_rank)
{
    // Initialize data_ as 15x3 matrix (flattened)
    data_.resize(15 * 3, 0.0f);
    ctrldata_.resize(1, 0);
}

void CoordReceiverAlgorithm::receive_data(int iter)
{
    int flag;
    MPI_Status status;

    MPI_Iprobe(partner_rank_, static_cast<int>(MessageType::eDataMesg), comm_, &flag, &status);
    if (flag)
    {
        MPI_Recv(data_.data(), data_.size(), MPI_FLOAT,
                 partner_rank_, static_cast<int>(MessageType::eDataMesg),
                 comm_, &status);

        std::cout << "Md C++ code with rank " << rank_
                  << " received data with shape : [15, 3] (md iteration: "
                  << iter << ")" << std::endl;
    }
}

void CoordReceiverAlgorithm::receive_ctrl_msg(int iter)
{
    int flag;
    MPI_Status status;

    MPI_Iprobe(partner_rank_, static_cast<int>(MessageType::eCtrlMsg), comm_, &flag, &status);
    if (flag)
    {
        MPI_Recv(ctrldata_.data(), ctrldata_.size(), MPI_INT,
                 partner_rank_, static_cast<int>(MessageType::eCtrlMsg),
                 comm_, &status);

        std::cout << "Md C++ code with rank " << rank_
                  << " received ctrl msg : " << ctrldata_[0]
                  << " (iteration count: " << iter << ")" << std::endl;
    }
}

void CoordReceiverAlgorithm::run(int iter)
{
    if (partner_rank_ != -1)
    {
        receive_data(iter);
        receive_ctrl_msg(iter);
    }
}

void calculate_energies(float iter_delay)
{
    std::this_thread::sleep_for(std::chrono::milliseconds(
        static_cast<int>(iter_delay)));
}

// Configuration structure to hold parameters
struct Config
{
    std::vector<std::string> md_lambda_values;
    int md_num_iter;
    float md_iter_delay;
};

std::vector<float> parseFloatVector(const std::string& str) {
    // Check for valid input format: starts with '[' and ends with ']'
    if (str.empty() || str.front() != '[' || str.back() != ']') {
        throw std::invalid_argument("Input string must start with '[' and end with ']'");
    }

    std::vector<float> vec;
    std::stringstream ss(str.substr(1, str.size() - 2));  // Remove square brackets and use stringstream

    for (std::string temp; std::getline(ss, temp, ','); ) {
        // Convert each trimmed string to float and add to vector
        vec.push_back(std::stof(temp.erase(0, temp.find_first_not_of(" \t")).erase(temp.find_last_not_of(" \t") + 1)));
    }

    return vec;
}

class ConfigPrototype {
public:
    ConfigPrototype(const std::string& file_name = "config.yaml")
    {
        /*YAML::Node config = YAML::LoadFile(file_name);
        md_num_iter = config["md_num_iter"].as<int>(3);
        md_iter_delay = config["receiving_interval"].as<float>(0.1)*1000;
        md_lambda_values = config["md_lambda_values"].as<std::vector<float>>(std::vector<float>());
        */
        Yaml::Node root;
        Yaml::Parse(root, file_name.c_str());
        md_num_iter = root["md_num_iter"].As<int>();
        md_iter_delay = root["md_iter_delay"].As<float>();
        md_lambda_values = parseFloatVector(root["md_lambda_values"].As<std::string>());
        /*
        if (lamdas.IsSequence())
        {
            for (auto i = 0 ; i != lamdas.Size();i++)
            {
                md_lambda_values.push_back(lamdas[i].As<float>());
            }
        }
        */
        md_num_iter = 3;
        md_iter_delay = 0.1 * 1000;
        md_lambda_values = {0,0.1,0.2,0.4};
    }
public:    
    int md_num_iter;
    float md_iter_delay;
    std::vector<float> md_lambda_values;

};

int main(int argc, char **argv)
{
    MPI_Init(&argc, &argv);

    MPI_Comm comm = MPI_COMM_WORLD;
    int rank;
    MPI_Comm_rank(comm, &rank);

    std::cout << "Receiver with rank " << rank << " started" << std::endl;

    auto config = ConfigPrototype();
    int partner_rank = sync_process(config.md_lambda_values, 'R', comm);

    CoordReceiverAlgorithm receiver(comm, rank, partner_rank);

    for (int iter = 0; iter < config.md_num_iter; ++iter)
    {
        calculate_energies(config.md_iter_delay);
        receiver.run(iter);
    }

    std::cout << "Receiver with rank " << rank << " finished" << std::endl;

    MPI_Finalize();
    return 0;
}

File: md_prototype.py
from mpi4py import MPI
import numpy as np
import time
from config_prototype import yaml_to_object
from mpi_helper import MessageType,CtrlMsg,sync_process

class CoordReceiverAlgorithm:  

    def __init__(self, comm, rank, partner_rank):
        self.comm = comm
        self.rank = rank
        self.partner_rank = partner_rank
        # Create a numpy array to receive into (like int in C)
        # data = np.empty(1, dtype='i')  # 'i' = 32-bit int
        self.data = np.zeros((15, 3), dtype=np.float32)
        self.ctrldata = np.zeros(1, dtype=np.int32)
    def receive_data(self,  iter):
        flag = self.comm.Iprobe(source=self.partner_rank, tag=MessageType.eDataMesg.value)
        if flag:
            self.comm.Recv(self.data, source=self.partner_rank, tag=MessageType.eDataMesg.value)
            print(
                f"Md Pythoncode with rank {self.rank} received data with shape : {self.data.shape} (md iterration: {iter})"
            )
        #energy1, energy2 = calclulate_energies(data)

    def receive_ctrl_msg(self, iter):
        flag = self.comm.Iprobe(source=self.partner_rank, tag=MessageType.eCtrlMsg.value)
        if flag:
            self.comm.Recv(self.ctrldata, source=self.partner_rank, tag=MessageType.eCtrlMsg.value)
            print(
                f"Md code with rank {self.rank} received ctrl msg : {self.ctrldata} (iterration count: {iter})"
            )

    def run(self, iter):
        if self.partner_rank != -1:
            self.receive_data(iter)
            self.receive_ctrl_msg(iter)
        
            

def calculate_energies(iter_delay):
    time.sleep(iter_delay)

def main():
    #config = ConfigPrototype()
    config = yaml_to_object()
    comm = MPI.COMM_WORLD
    rank = comm.Get_rank()
    print(f"Receiver with rank {rank} started")
    partner_rank = sync_process(config.md_lambda_values, 'R', comm)
    receiver = CoordReceiverAlgorithm(comm, rank, partner_rank)
    for iter in range(config.md_num_iter):
        calculate_energies(config.md_iter_delay)
        receiver.run(iter)
            
    print(f"Receiver with rank {rank} finished")

main()
